flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageMemoryManagerImpl.java

```diff
@@ -0,0 +1,279 @@
+     */
+    private boolean isInitialized;
+
+    /**
+     * The constructor of the {@link TieredStorageMemoryManagerImpl}.
+     *
+     * @param numTriggerReclaimBuffersRatio the buffer pool usage ratio of requesting each tiered
+     *     storage to reclaim buffers
+     * @param needPeriodicalCheckReclaimBuffer indicate whether it is necessary to start a
+     *     periodically checking buffer reclaim thread
+     */
+    public TieredStorageMemoryManagerImpl(
+            float numTriggerReclaimBuffersRatio, boolean needPeriodicalCheckReclaimBuffer) {
+        this.numTriggerReclaimBuffersRatio = numTriggerReclaimBuffersRatio;
+        this.needPeriodicalCheckReclaimBuffer = needPeriodicalCheckReclaimBuffer;
+        this.tieredMemorySpecs = new HashMap<>();
+        this.numRequestedBuffers = new AtomicInteger(0);
+        this.bufferReclaimRequestListeners = new ArrayList<>();
+        this.isInitialized = false;
+    }
+
+    @Override
+    public void setup(BufferPool bufferPool, List<TieredStorageMemorySpec> storageMemorySpecs) {
+        this.bufferPool = bufferPool;
+        for (TieredStorageMemorySpec memorySpec : storageMemorySpecs) {
+            checkState(
+                    !tieredMemorySpecs.containsKey(memorySpec.getOwner()),
+                    "Duplicated memory spec.");
+            tieredMemorySpecs.put(memorySpec.getOwner(), memorySpec);
+            numTotalGuaranteedBuffers += memorySpec.getNumGuaranteedBuffers();
+        }
+
+        if (needPeriodicalCheckReclaimBuffer) {
+            this.executor =
+                    Executors.newSingleThreadScheduledExecutor(
+                            new ThreadFactoryBuilder()
+                                    .setNameFormat("buffer reclaim checker")
+                                    .setUncaughtExceptionHandler(FatalExitExceptionHandler.INSTANCE)
+                                    .build());
+        }
+
+        this.isInitialized = true;
+    }
+
+    @Override
+    public void listenBufferReclaimRequest(Runnable onBufferReclaimRequest) {
+        bufferReclaimRequestListeners.add(onBufferReclaimRequest);
+    }
+
+    /**
+     * Request a {@link BufferBuilder} instance from {@link BufferPool} for a specific owner. The
+     * {@link TieredStorageMemoryManagerImpl} will not check whether a buffer can be requested and
+     * only record the total number of requested buffers. If the buffers in the {@link BufferPool}
+     * is not enough, this will request each tiered storage to reclaim the buffers as much as
+     * possible.
+     *
+     * <p>Note that synchronization is not necessary for this method, as it can only be called by
+     * the task thread.
+     *
+     * @return the requested buffer
+     */
+    @Override
+    public BufferBuilder requestBufferBlocking() {
+        checkIsInitialized();
+
+        tryReclaimBuffers();
+        CompletableFuture<MemorySegment> requestedBuffer =
+                CompletableFuture.supplyAsync(
```

Comment:
This involves an unnecessary thread. Creating an future and completes it after `bufferPool.requestMemorySegmentBlocking()` should be good enough.

Reply:
Good catch. Fixed by a `CompletableFuture<Void>`.